torch_geometric/nn/models/pg_explainer.py

```diff
@@ -0,0 +1,270 @@
+from typing import Optional
+
+import torch
+from torch import Tensor, nn
+from torch.nn.functional import nll_loss
+from torch_scatter import scatter_mean
+from tqdm import tqdm
+
+from torch_geometric.nn.models.explainer import (Explainer, clear_masks,
+                                                 set_masks)
+
+EPS = 1e-15
+
+
+class PGExplainer(Explainer):
+    r"""The PGExplainer model from the `"Parameterized Explainer for Graph Neural
+    Network" <https://arxiv.org/abs/2011.04573>`_ paper. It uses a neural
+    network :obj:`explainer_model` , to predict which edges are crucial to
+    a GNNs node or graph prediction.
+
+    .. note::
+
+        For an example of using PGExplainer, see `examples/pg_explainer.py
+        <https://github.com/pyg-team/pytorch_geometric/blob/master/examples/
+        pg_explainer.py>`_.
+
+    Args:
+        model (torch.nn.Module): The GNN module to explain.
+        out_channels(int): Size of output of the last :class:`MessagePassing`
+            layer in :obj:`model`.
+        epochs (int, optional): The number of epochs to train.
+            (default: :obj:`30`)
+        lr (float, optional): The learning rate to apply.
+            (default: :obj:`0.003`)
+        num_hops (int, optional): The number of hops the :obj:`model` is
+            aggregating information from.
+            If set to :obj:`None`, will automatically try to detect this
+            information based on the number of
+            :class:`~torch_geometric.nn.conv.message_passing.MessagePassing`
+            layers inside :obj:`model`. (default: :obj:`None`)
+        task (str): Denotes the type of task that needs explanation. Valid
```

Comment:
This is probably something we can address in another PR.
`PGExplainer` and `GNNExplainer` have a different way of choosing graph and node tasks. `GNNExplainer` uses different functions and `PGExplianer` uses an if condition.

Reply:
I guess that a single function `explain` (and `train_explainer`) is more practical. However, in this case, we would have to change the API for the GNNExplainer class. What do you think is the better option here? I also see you already considered this when introducing the graph explanation method for `GNNExplainer`.

Reply:
Yeah two functions `explain` and `train_explainer` (GNNExplainer wouldn't have this) seems cleaner to meet too. Although this would be a breaking change to `GNNExplainer`. @rusty1s  WDYT?

Reply:
Ah nice idea @rusty1s .